### PR TITLE
net: http: Provide destination address in http replies

### PR DESCRIPTION
**Pull request by rveerama1**
net_app_ctx maintains multiple net contexts(net_ctx). But when http
api's wants to reply or send some data, its always choose the first
net_context in the array, which is not correct always.

net_app_get_net_pkt_with_dst() api will select proper context
based on destination address. So with the help of new api in
net_app, http can select proper context and send packets. To
achieve this, desination address is provided in http_recv_cb_t
and http_connect_cb_t callbacks. Also chaged relevant API's to
provide destination address in http message preparation methods.

### CODE REVIEW

**Comment by mbolivar** [comment]
Do you have a summary description for what changes existing applications must make in the meantime?

**Comment by mike-scott** [comment]
@rveerama1 This patch is semi confusing.  The change from net_app_get_net_pkt() -> net_app_get_net_pkt_with_dst() will make sure you select a context w/ the same family as dst, but it doesn't actually check any other information.  Is that the goal of the PR?  If so, why use sockaddr?  Would passing sa_family around as the parameter be clearer?

**Comment by rveerama1** [comment]
> Do you have a summary description for what changes existing applications must make in the meantime?

I am working on changing samples and tests. But in mean time take a look 
https://github.com/zephyrproject-rtos/zephyr/pull/5035/commits/1f838a5df52df0ce9d725c63b48a090ce5d77006#diff-780b2211d7789425a2584af4afc52b77 . This should give an overview of changes to application.

**Comment by rveerama1** [comment]
> @rveerama1 This patch is semi confusing. The change from net_app_get_net_pkt() -> net_app_get_net_pkt_with_dst() will make sure you select a context w/ the same family as dst, but it doesn't actually check any other information. Is that the goal of the PR? If so, why use sockaddr? Would passing sa_family around as the parameter be clearer?

@mike-scott https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/app/net_app.c#L966 and https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/app/net_app.c#L656. In case of TCP select_server_ctx() calls get_server_ctx() should match dst address. But in case of UDP https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/app/net_app.c#L665 you are assumption is right. Issue might exist in case UDP when net_app_ctx has multiple net_contexts and when selection proper context based on dst address may go wrong. I don't have any samples to verify. Have you tried anytime with net_app_ctx with multiple active net_contexts (in case of UDP) and send data over different contexts?

**Comment by rveerama1** [comment]
> Issue might exist in case UDP when net_app_ctx has multiple net_contexts and when selection proper context based on dst address may go wrong. I don't have any samples to verify. Have you tried anytime with net_app_ctx with multiple active net_contexts (in case of UDP) and send data over different contexts?

Bit thorough looking at net_app.h only TCP in case of Server role has multiple net contexts. UDP shouldn't be having these kind of issue.

**Comment by rveerama1** [comment]
Tests doesn't need any changes and samples are modified

**Comment by jukkar** [comment]
@mbolivar Could you test with this patch and latest upstream if you can replicate the issue you were seeing earlier?

**Comment by mbolivar** [comment]
> @mbolivar Could you test with this patch and latest upstream if you can replicate the issue you were seeing earlier?
> 

@jukkar @rveerama1 this is helpful, but not a complete fix, for the issue we are seeing. With this patch, we no longer see usage faults, but our application is still broken by what looks like a spurious HTTP timeout.

I'm trying to make a smaller reproducer.

**Review by mike-scott — DISMISSED** [review]
This patch + https://github.com/zephyrproject-rtos/zephyr/pull/6410 fixes our HTTP-based sample.

**Comment by mike-scott** [comment]
@rveerama1 please note that this patch re-adds a patch that was previously removed during the API cleanup: https://github.com/zephyrproject-rtos/zephyr/commit/e04a5412a12f6b18b8f142aa488f75f1f81aa93f

I've re-added that patch to my patches here: https://github.com/zephyrproject-rtos/zephyr/pull/6410

This will of course cause this PR to have a conflict.  Hopefully it can be fixed easily if my PR gets merged first?

**Review by mike-scott — COMMENTED** [review]
Will need rebase when https://github.com/zephyrproject-rtos/zephyr/pull/6410 gets merged.

**Comment by rveerama1** [comment]
> Will need rebase when #6410 gets merged.

I will rebase.

**Comment by carlescufi** [comment]
#6410 is merged, if you want this to go into 1.11 please review ASAP

**Comment by rveerama1** [comment]
@carlescufi Rebased and uploaded.
@jukkar @mike-scott @mbolivar can you take a look please.

File: subsys/net/lib/http/http_server.c [hunk 1]
```diff
@@ -566,6 +572,11 @@ static void http_received(struct net_app_ctx *app_ctx,
 
 	NET_DBG("[%p] Received %zd bytes http data", ctx, recv_len);
 
+	if (net_pkt_context(pkt)) {
+		ctx->http.parser.addr = &net_pkt_context(pkt)->remote;
```
Comment (mike-scott) [inline subsys/net/lib/http/http_server.c]: Does ctx->http.parser.addr ever get used anywhere?  I can't find it in this patch.

File: include/net/http.h [hunk 2]
```diff
@@ -445,12 +451,6 @@ struct http_ctx {
 	/** Network buffer allocation timeout */
 	s32_t timeout;
 
-	/** Websocket endpoint address */
-	struct sockaddr *addr;
-
-	/** Websocket endpoint port */
-	u16_t port;
```
Comment (mike-scott) [inline include/net/http.h]: Was the port member not being used?

File: subsys/net/lib/http/http_server.c [hunk 3]
```diff
@@ -311,7 +313,7 @@ static inline void new_client(struct http_ctx *ctx,
 	struct net_context *net_ctx;
 	const char *type_str = "HTTP";
 
-	net_ctx = get_server_ctx(app_ctx, ctx->addr);
+	net_ctx = get_server_ctx(app_ctx, NULL);
```
Comment (mike-scott) [inline subsys/net/lib/http/http_server.c]: This will always return NULL now.  See get_server_ctx() in http_server.c

**Review by mike-scott — CHANGES_REQUESTED** [review]

File: subsys/net/lib/http/http_server.c (hunk 1 above)
Comment (rveerama1) [reply subsys/net/lib/http/http_server.c]: It's used in websocket reply. https://github.com/zephyrproject-rtos/zephyr/pull/5035/commits/53eb1b6eab6d1f825c98e761b2e5995f767ed5a4#diff-ee0f87c7adaeba40cb43dd5749f9407aR354

File: include/net/http.h (hunk 2 above)
Comment (rveerama1) [reply include/net/http.h]: Yep, Never used.

File: subsys/net/lib/http/http_server.c (hunk 3 above)
Comment (rveerama1) [reply subsys/net/lib/http/http_server.c]: True, I will fix this.

**Comment by rveerama1** [comment]
addressed @mike-scott comments. Uploaded new version.

**Review by mike-scott — APPROVED** [review]
I didn't test latest version of the patch.  But code LGTM.

**Comment by rveerama1** [comment]
> I didn't test latest version of the patch. But code LGTM.

Not many changes in this patch, only dst addr provided as per your comment, that too it's a debug function.

**Comment by mike-scott** [comment]
> Not many changes in this version, only dst addr provided as per your comment, that too it's a debug function.

Yep.  I don't think anything here will break us.  (I approved the PR).

**Comment by carlescufi** [comment]
@jukkar @tbursztyka I would need an additional review to merge this. @rveerama1 is this critical for 1.11?

**Review by jukkar — APPROVED** [review]
+1 from me

**Comment by rveerama1** [comment]
> @rveerama1 is this critical for 1.11?

Yes, if Zephyr is acting as a HTTP server with concurrent TCP connections and multiple active net_contexts, without this PR applications will fail to send data to right peer address.

**Pull request merged**